batcher/batcher.go

```diff
@@ -158,6 +165,11 @@ func (b *basicBatcher) Dispose() {
 	b.flush()
 	b.disposed = true
 	b.items = nil
+
+	// Drain the batch channel and all routines waiting to put on the channel
+	for len(b.batchChan) > 0 || atomic.LoadInt32(&b.waiting) > 0 {
+		<-b.batchChan
+	}
 	close(b.batchChan)
```

Comment:
Each routine bumps ```b.waiting``` to ensure the drain runs enough times. See [here](https://github.com/Workiva/go-datastructures/blob/batcher_deadlock/batcher/batcher.go#L195).